<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bayesianische Variablenselektion • bayeskurs</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bayeskurs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Bayesianische Variablenselektion</h1>
                        <h4 class="author">Volker Schmid</h4>
            
            <h4 class="date"><ol start="10" style="list-style-type: decimal">
<li>Juli 2017</li>
</ol></h4>
          </div>

    
    
<div class="contents">
<div id="ridge-und-lasso" class="section level1">
<h1 class="hasAnchor">
<a href="#ridge-und-lasso" class="anchor"></a>Ridge und Lasso</h1>
<div id="bayesianisches-generalisiertes-lineares-modell" class="section level2">
<h2 class="hasAnchor">
<a href="#bayesianisches-generalisiertes-lineares-modell" class="anchor"></a>Bayesianisches (Generalisiertes) Lineares Modell</h2>
<p>Gegeben seien <span class="math inline">\(n\)</span> Beobachtungen einer Zielvariable <span class="math inline">\(y\)</span> und von <span class="math inline">\(p\)</span> Kovariablen <span class="math inline">\(x_1,\ldots,x_p\)</span>.</p>
<p><span class="math display">\[ \begin{aligned}
y_i|\mu_i,\phi_i &amp;\sim f(\mu_i,\phi_i) \text{ i.i.d.}\\
h(\mu_i) &amp; = \sum_{j=1}^p \beta_j x_{ij} 
\end{aligned}\]</span></p>
</div>
<div id="bayesianisches-lineares-regressionsmodell" class="section level2">
<h2 class="hasAnchor">
<a href="#bayesianisches-lineares-regressionsmodell" class="anchor"></a>Bayesianisches lineares Regressionsmodell</h2>
<p><span class="math display">\[\begin{aligned}
y_i|\beta,\sigma^2 &amp; \sim {\rm N}(\mu_i,\sigma^2)\\
\mu_i &amp; = \sum_{j=1}^p \beta_j x_{ij}
\end{aligned}\]</span></p>
<p>Welche Priori-Information haben wir über die <span class="math inline">\(\beta\)</span>? Erstmal keine…</p>
<p><span class="math display">\[p(\beta_j) \propto \text{const.} \]</span> Können wir sehen als uneigentliche Normalverteilung (hier konjugierte Verteilung) mit Varianz unendlich. Damit ist</p>
<p><span class="math display">\[ \beta\sim N_p(\hat{\beta},\Sigma)\]</span></p>
<p>mit <span class="math inline">\(\hat{\beta} = (X'X)^{-1}X'y\)</span> dem KQ-Schätzer!</p>
</div>
<div id="ridge-regression" class="section level2 allowframebreaks">
<h2 class="hasAnchor">
<a href="#ridge-regression" class="anchor"></a>Ridge-Regression</h2>
<p>Nun: <span class="math inline">\(p&gt;&gt;n\)</span>.</p>
<p>Idee der Ridge-Regression: Viele der <span class="math inline">\(\beta\)</span> Parameter sollen gleich oder nahe Null sein. Bestrafe daher Parameter, die zu stark von der Null abweichen. Damit penalisierter log-Likelihood-Ansatz:</p>
<p><span class="math display">\[ l_{pen}(\beta)=l(\beta)-\frac{\lambda}{2} \sum_{j=1}^p \beta_j^2 \]</span></p>
<p>Bayesianisch gedacht: Wir haben die Vorinformation, dass die Parameter nahe Null sind. Kombiniert mit konjugiertem Priori-Ansatz kommen wir auf:</p>
<p><span class="math display">\[ \beta_j \sim N(0,\tau^{-1}) \quad \forall j \]</span></p>
<p>Die Log-Priori-Dichte ist <span class="math display">\[ \log(p(\beta))=-\frac{\tau}{2}\sum_{j=1}^p \beta^2 +C \]</span></p>
<p>Damit sind penalisierte log-Likelihood und log-Posteriori bis auf Konstanten identisch. Ein Maximum-A-Posteriori-Ansatz liefert also selbes Ergebnis wie ein penalisierter log-Likelihood-Ansatz (Tikhonov-Regularisierung).</p>
</div>
<div id="beispiel-bayesianische-ridge-regression" class="section level2 allowframebreaks">
<h2 class="hasAnchor">
<a href="#beispiel-bayesianische-ridge-regression" class="anchor"></a>Beispiel Bayesianische Ridge-Regression</h2>
<p>Wir konstruieren einen Beispieldatensatz</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="dv">50</span> 
p &lt;-<span class="st"> </span><span class="dv">100</span>
true.sigma2 &lt;-<span class="st"> </span><span class="fl">0.001</span>
x &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">runif</span>(n*p), <span class="dt">nrow=</span>n)
true.beta &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">20</span>,<span class="dv">30</span>, <span class="kw">rep</span>(<span class="dv">0</span>,p<span class="dv">-3</span>))
mu &lt;-<span class="st"> </span><span class="kw">as.vector</span>(x%*%true.beta)
y &lt;-<span class="st"> </span><span class="kw">rnorm</span>(n,mu,<span class="kw">sqrt</span>(true.sigma2))
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))
for (i in <span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">10</span>,<span class="dv">90</span>))
  <span class="kw">plot</span>(x[,i],y)</code></pre></div>
<p><img src="Variablenselektion_files/figure-html/ridge-daten-1.png" width="672"></p>
</div>
<div id="posteriori" class="section level2">
<h2 class="hasAnchor">
<a href="#posteriori" class="anchor"></a>Posteriori</h2>
<p><span class="math display">\[ \begin{aligned}
p(\beta,\tau,\sigma^2|y) &amp;\propto \sigma^{-n}\exp\left(-\frac{1}{2\sigma^2}\sum_i(y_i-\sum_j\beta_jx_{ij})^2\right)\\
&amp;\cdot \tau^{p/2}\exp\left(-\frac{\tau}{2}\sum_j\beta_j^2\right)\\
&amp; \cdot \tau^{a-1}\exp(-\tau b)\\
&amp; \cdot \sigma^2{-a_0-1}\exp(-b_0/\sigma^2)\\
\end{aligned} \]</span></p>
<p>Damit gilt:</p>
<ul>
<li>
<span class="math inline">\(\beta|\tau,\sigma^2 \sim N(\hat{\beta},\Sigma)\)</span> mit <span class="math inline">\(\hat{\beta}=(X'X+\tau I)^{-1}X'y\)</span> und <span class="math inline">\(\Sigma=(X'X+\tau I)^{-1})\)</span>.</li>
<li><span class="math inline">\(\tau|\beta \sim Ga(a+p/2,b+\sum\beta_j^2/2)\)</span></li>
<li>
<span class="math inline">\(\sigma^2|\beta,y \sim IG(a_0+n/2,b+\sum(\epsilon_i^2))\)</span> mit <span class="math inline">\(\epsilon_i=y_i-\sum_j\beta_jx_{ij}\)</span>
</li>
</ul>
</div>
<div id="mmcm" class="section level2 allowframebreaks">
<h2 class="hasAnchor">
<a href="#mmcm" class="anchor"></a>MMCM</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">beta&lt;-<span class="kw">rep</span>(<span class="dv">0</span>,p)
XX &lt;-<span class="st"> </span><span class="kw">t</span>(x)%*%x
Xy &lt;-<span class="st"> </span><span class="kw">t</span>(x)%*%y
tau &lt;-<span class="st"> </span><span class="dv">1</span>
sigma2 &lt;-<span class="st"> </span><span class="dv">1</span>
a0 &lt;-<span class="st"> </span><span class="dv">1</span>
b0 &lt;-<span class="st"> </span><span class="fl">0.001</span>
a &lt;-<span class="st"> </span><span class="dv">1</span>
b &lt;-<span class="st"> </span><span class="fl">0.1</span>

beta.save&lt;-<span class="kw">array</span>(<span class="ot">NA</span>,<span class="kw">c</span>(p,<span class="dv">500</span>))
tau.save&lt;-<span class="kw">rep</span>(<span class="ot">NA</span>,<span class="dv">500</span>)
sigma2.save&lt;-<span class="kw">rep</span>(<span class="ot">NA</span>,<span class="dv">500</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">for (i in <span class="dv">1</span>:<span class="dv">1000</span>)
{
  Sigma &lt;-<span class="st"> </span><span class="kw">solve</span>(XX+tau*<span class="kw">diag</span>(p))
  mu &lt;-<span class="st"> </span>Sigma%*%Xy
  beta &lt;-<span class="st"> </span>mnormt::<span class="kw">rmnorm</span>(<span class="dv">1</span>,mu,Sigma)
  
  tau &lt;-<span class="st"> </span><span class="kw">rgamma</span>(<span class="dv">1</span>, a+p/<span class="dv">2</span>, b+<span class="kw">sum</span>(beta^<span class="dv">2</span>)/<span class="dv">2</span>)

  sigma2 &lt;-<span class="st"> </span><span class="dv">1</span>/<span class="kw">rgamma</span>(<span class="dv">1</span>, a0+n/<span class="dv">2</span>, b0+<span class="kw">sum</span>((y-x%*%beta)^<span class="dv">2</span>))
  
  if (i&gt;<span class="dv">500</span>)
  {
    beta.save[,i<span class="dv">-500</span>]=beta
    tau.save[i<span class="dv">-500</span>]=tau
    sigma2.save[i<span class="dv">-500</span>]=sigma2
  }
}</code></pre></div>
</div>
<div id="plot" class="section level2">
<h2 class="hasAnchor">
<a href="#plot" class="anchor"></a>Plot</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">beta.qu&lt;-<span class="kw">apply</span>(beta.save,<span class="dv">1</span>,quantile,<span class="dt">probs=</span><span class="kw">c</span>(.<span class="dv">05</span>,.<span class="dv">5</span>,.<span class="dv">95</span>))
<span class="kw">plot</span>(beta.qu[<span class="dv">2</span>,],<span class="dt">pch=</span><span class="dv">19</span>,<span class="dt">ylim=</span><span class="kw">range</span>(beta.qu),<span class="dt">ylab=</span><span class="st">"beta"</span>)
for (i in <span class="dv">1</span>:p)
<span class="kw">lines</span>(<span class="kw">rep</span>(i,<span class="dv">2</span>),beta.qu[<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>),i],<span class="dt">lwd=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="Variablenselektion_files/figure-html/ridge-plot-1.png" width="672"></p>
</div>
<div id="relevante-kovariablen" class="section level2">
<h2 class="hasAnchor">
<a href="#relevante-kovariablen" class="anchor"></a>Relevante Kovariablen</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(beta.qu[<span class="dv">2</span>,]==<span class="dv">0</span>)</code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">which</span>(beta.qu[<span class="dv">1</span>,]&gt;<span class="dv">0</span>))</code></pre></div>
<pre><code>## [1]  1  2  3 87</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">which</span>(beta.qu[<span class="dv">3</span>,]&lt;<span class="dv">0</span>))</code></pre></div>
<pre><code>## integer(0)</code></pre>
<ul>
<li>Bei Ridge werden die Parameter Richtung Null gedrückt</li>
<li>Aber: Parameter werden nicht genau gleich Null!</li>
</ul>
</div>
<div id="lasso" class="section level2">
<h2 class="hasAnchor">
<a href="#lasso" class="anchor"></a>Lasso</h2>
<p>Alternative: Lasso (<span class="math inline">\(L_1\)</span>-Regularisierung)</p>
<p><span class="math display">\[ pen(\beta)=\sum_j|\beta_j| \]</span></p>
<p>Bayesianisch analog zu Ridge:</p>
<p><span class="math display">\[ p(\beta_j) \propto \exp\left(-\frac{\tau}{2} \sum |\beta_j| \right)\]</span></p>
<ul>
<li>Das entspricht einer Laplace-Verteilung mit Erwartungswert 0</li>
<li>Aber: (erstmal) kein Gibbs-Sampler mehr möglich</li>
</ul>
</div>
<div id="bayesianischer-lasso" class="section level2 allowframebreaks">
<h2 class="hasAnchor">
<a href="#bayesianischer-lasso" class="anchor"></a>Bayesianischer Lasso</h2>
<p>Nach <em>Park, Trevor and Casella, George. The Bayesian Lasso. Journal of American Statistical Association. 103(482):681-686. 2008</em> gilt äquivalent:</p>
<p><span class="math display">\[ \begin{aligned}
\beta_j | \sigma^2,\tau_j^2 &amp;\sim N(0,\sigma^2\tau_j^2)\\
\tau_j|\sigma^2 &amp; \sim Exp(\lambda^2/2)\\
\lambda^2 &amp;\sim Ga(a,b)
\end{aligned} \]</span></p>
<p>Damit läßt sich wiederum ein Gibbs-Sampler konstruieren.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">beta.L&lt;-<span class="kw"><a href="../reference/gibbsBLasso.html">gibbsBLasso</a></span>(x, y, <span class="dt">max.steps =</span> <span class="dv">10000</span>)</code></pre></div>
<pre><code>## Iteration: 1000 
Iteration: 2000 
Iteration: 3000 
Iteration: 4000 
Iteration: 5000 
Iteration: 6000 
Iteration: 7000 
Iteration: 8000 
Iteration: 9000 
Iteration: 10000 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(beta.L[<span class="dv">2</span>,],<span class="dt">pch=</span><span class="dv">19</span>,<span class="dt">ylim=</span><span class="kw">range</span>(beta.L),<span class="dt">ylab=</span><span class="st">"beta"</span>)
for (i in <span class="dv">1</span>:p)
<span class="kw">lines</span>(<span class="kw">rep</span>(i,<span class="dv">2</span>),beta.L[<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>),i],<span class="dt">lwd=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="Variablenselektion_files/figure-html/lasso-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(beta.L[<span class="dv">2</span>,]==<span class="dv">0</span>)</code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">which</span>(beta.L[<span class="dv">1</span>,]&gt;<span class="dv">0</span>))</code></pre></div>
<pre><code>##  [1]   1   2   3   5  11  17  24  25  29  33  36  37  43  44  45  47  59
## [18]  60  61  63  73  81  82  83  84  85  90  94 100</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">which</span>(beta.L[<span class="dv">3</span>,]&lt;<span class="dv">0</span>))</code></pre></div>
<pre><code>## integer(0)</code></pre>
</div>
<div id="elastic-net" class="section level2">
<h2 class="hasAnchor">
<a href="#elastic-net" class="anchor"></a>Elastic Net</h2>
<p>Ridge und Lasso lassen sich kombinieren:</p>
<p><span class="math display">\[ p(\beta_j) \propto \exp\left(-\frac{\tau}{2} \sum |\beta_j| -\frac{\nu}{2} \sum \beta_j^2 \right)\]</span></p>
<p><img src="Variablenselektion_files/figure-html/priorplot-1.png" width="672"></p>
</div>
</div>
<div id="indikatorvariablen" class="section level1">
<h1 class="hasAnchor">
<a href="#indikatorvariablen" class="anchor"></a>Indikatorvariablen</h1>
<div id="indikatorvariablen-1" class="section level2">
<h2 class="hasAnchor">
<a href="#indikatorvariablen-1" class="anchor"></a>Indikatorvariablen</h2>
<p>Setze</p>
<p><span class="math display">\[\beta_i = I_i\tilde{\beta}_i \]</span></p>
<p>wobei <span class="math inline">\(I_i\)</span> eine (0/1-)Indikatorvariable ist.</p>
<p>Ist <span class="math inline">\(I_i=0\)</span>, wird <span class="math inline">\(\beta_i\)</span> auf <span class="math inline">\(0\)</span> gesetzt, <span class="math inline">\(\tilde{\beta}_i\)</span> wird aus der Priori gezogen.</p>
</div>
<div id="ising-feld" class="section level2">
<h2 class="hasAnchor">
<a href="#ising-feld" class="anchor"></a>Ising-Feld</h2>
<ul>
<li>Ansatz lässt sich auch auf Kovariablen mit bekannter/angenommener Korrelation anwenden</li>
<li>Z.B. Gene auf DNA, Bilder</li>
<li>Auf die Indikatorvariablen wird das ein Ising-Feld angenommen</li>
<li>mit <span class="math inline">\(J_i=2I_i-1\)</span>, also <span class="math inline">\(J_i\in\{-1,1\}\)</span>
</li>
</ul>
<p><span class="math display">\[ p(I) \propto \exp\left(-\tau \sum_{i\sim j}J_i J_j\right) \]</span></p>
<ul>
<li>Sampling daraus allerdings schwierig</li>
<li>Alternative: Probit-Modell</li>
</ul>
<p><span class="math display">\[ I_i = \left\{ \begin{aligned}
1 &amp; \text{ für } \phi&gt;0\\
0 &amp; \text{ für } \phi\leq 0 \end{aligned}\right. 
\]</span></p>
<p>mit</p>
<p><span class="math display">\[p(\phi) \propto \left(\exp(-\frac{\tau}{2}\sum_{i\sim j}(\phi_i-\phi_j)^2\right)\]</span></p>
</div>
</div>
<div id="spike-and-slab" class="section level1">
<h1 class="hasAnchor">
<a href="#spike-and-slab" class="anchor"></a>Spike and Slab</h1>
<div id="spike-and-slab-1" class="section level2 allowframebreaks">
<h2 class="hasAnchor">
<a href="#spike-and-slab-1" class="anchor"></a>Spike and Slab</h2>
<ul>
<li>Idee: Damit <span class="math inline">\(p(\beta=0|y)&gt;0\)</span>, muss <span class="math inline">\(p(\beta=0)&gt;0\)</span> sein</li>
<li>Kombiniere flache Priori (Slab) mit Punktmasse auf Null (Spike)<br>
</li>
<li>Computational bessere Darstellung als Mischung von zwei Normalverteilungen mit sehr großer uns sehr kleiner Varianz</li>
</ul>
<p><img src="Variablenselektion_files/figure-html/spikeslabplot-1.png" width="672"></p>
<ul>
<li>Änhlichkeiten zu Elastic Net, wenn Lasso über Normalverteilung modelliert wird.</li>
<li>z.B. Implementation in spikeSlabGAM-Paker</li>
</ul>
<p><span class="math display">\[ \begin{aligned}
\beta|\gamma,\tau^2 &amp;\sim N(0,\tau^2\gamma)\\
\gamma|w &amp; \sim wI_1(\gamma) + (1-w)I_{\nu_0}(\gamma)\\
\tau^2&amp;\sim IG(a_\tau, b_\tau)\\
w&amp;\sim Beta(a_w, b_w)
\end{aligned}\]</span></p>
<ul>
<li>
<span class="math inline">\(\nu_0\)</span> sehr klein, entspricht <em>spike</em>
</li>
</ul>
</div>
<div id="wgrr-und-gen" class="section level2">
<h2 class="hasAnchor">
<a href="#wgrr-und-gen" class="anchor"></a>WGRR und gen</h2>
<ul>
<li>Ishwaran und Rao (2014) zeigen, dass die Spike and Slab prior ein Spezialfall der gewichteten generalisierten Ridge-Regression sind (weighted generalized Ridge regression, WGRR)</li>
<li>Beim WGRR können die <span class="math inline">\(\beta\)</span> Parameter auf Null gesetzt werden, wir erhalten <span class="math inline">\(p(\beta=0|y)\)</span>
</li>
<li>Durch Bayesian Model Averaging gehen aber z.B. bei <span class="math inline">\(E(y^*|y)\)</span> weiterhin viele/alle Regressionsparameter ein</li>
<li>Ishwaran und Rao (2014) schlagen weiterhin das generalizes elatic net (gen) vor, das mehr Paramete auf Null setzt</li>
</ul>
</div>
<div id="spikeslabgam-paket" class="section level2 allowframebreaks">
<h2 class="hasAnchor">
<a href="#spikeslabgam-paket" class="anchor"></a>spikeslabGAM paket</h2>
<p>Beim spikeSlabGAM-Paket wird die Variablenselektion auf die glatten Effekte angewandt:</p>
<pre><code>## ## ----- This is spikeSlabGAM 1.1-11 ----- ##
## Please note that a recent update to gridExtra has made it necessary
##   to change the interface for &lt;plot.spikeSlabGAM&gt; starting in version 1.1-9.
## Instead of arguments 'rows', 'cols', 'widths', 'heights', 'maxPlotsPerPage',
##   it now accepts only 'nrow' and 'ncol'.
## Arguments 'widths' &amp; 'heights' can still be defined and are handed over
##   to &lt;gridExtra:::marrangeGrob&gt;.
## Sorry for the inconvenience.</code></pre>
<p><img src="Variablenselektion_files/figure-html/spikeslabgam-data-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mcmc &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">nChains=</span><span class="dv">4</span>, <span class="dt">chainLength=</span><span class="dv">1000</span>, <span class="dt">burnin=</span><span class="dv">500</span>, <span class="dt">thin=</span><span class="dv">5</span>)
m0 &lt;-<span class="st"> </span><span class="kw">spikeSlabGAM</span>(diabetes ~<span class="st"> </span>pregnant +<span class="st"> </span>glucose +<span class="st"> </span>pressure +<span class="st"> </span>mass +<span class="st"> </span>pedigree +<span class="st"> </span>age,
        <span class="dt">family=</span><span class="st">"binomial"</span>, <span class="dt">data=</span>pimaDiabTrain, <span class="dt">mcmc=</span>mcmc)</code></pre></div>
<pre><code>## 
## Model has  58  coefficients in  13 model terms.
## Blockwise sampling: alpha: 3  block(s); xi: 5 block(s).</code></pre>
<pre><code>## Using 2 parallel processes.
## Use 'options(mc.cores = &lt;YourNumberHere&gt;)' to override next time.</code></pre>
<pre><code>## 
## starting chain(s):
## bbbb0--------------------------------------100%
## 
## Mean acceptance rates:
## alpha   ksi 
##  0.93  0.63</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">summary</span>(m0), <span class="dt">printModels=</span><span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## Spike-and-Slab STAR for Binomial data 
## 
## Model:
## diabetes ~ (lin(pregnant) + sm(pregnant)) + (lin(glucose) + sm(glucose)) + 
##     (lin(pressure) + sm(pressure)) + (lin(mass) + sm(mass)) + 
##     (lin(pedigree) + sm(pedigree)) + (lin(age) + sm(age))
## 524 observations; 58 coefficients in 13 model terms.
## 
## Prior:
##  a[tau]  b[tau]    v[0]    a[w]    b[w] 
## 5.0e+00 2.5e+01 2.5e-04 1.0e+00 1.0e+00 
## 
## MCMC:
## Saved 4000 samples from 4 chain(s), each ran 5000 iterations after a
##   burn-in of 500 ; Thinning: 5
## P-IWLS acceptance rates: 0.93 for alpha; 0.63 for xi.
##                             
## Null deviance:           674
## Mean posterior deviance: 448
## 
## Marginal posterior inclusion probabilities and term importance:
##               P(gamma = 1)     pi dim    
## u                       NA     NA   1    
## lin(pregnant)        0.688  0.057   1  **
## sm(pregnant)         0.196  0.001   8    
## lin(glucose)         1.000  0.640   1 ***
## sm(glucose)          0.013  0.000   9    
## lin(pressure)        0.015  0.000   1    
## sm(pressure)         0.018  0.000   9    
## lin(mass)            1.000  0.187   1 ***
## sm(mass)             0.220  0.009   9    
## lin(pedigree)        0.349  0.011   1   *
## sm(pedigree)         0.229 -0.002   8    
## lin(age)             0.269  0.023   1   *
## sm(age)              1.000  0.074   8 ***
## *:P(gamma = 1)&gt;.25 **:P(gamma = 1)&gt;.5 ***:P(gamma = 1)&gt;.9</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#ridge-und-lasso">Ridge und Lasso</a><ul class="nav nav-pills nav-stacked">
<li><a href="#bayesianisches-generalisiertes-lineares-modell">Bayesianisches (Generalisiertes) Lineares Modell</a></li>
      <li><a href="#bayesianisches-lineares-regressionsmodell">Bayesianisches lineares Regressionsmodell</a></li>
      <li><a href="#ridge-regression">Ridge-Regression</a></li>
      <li><a href="#beispiel-bayesianische-ridge-regression">Beispiel Bayesianische Ridge-Regression</a></li>
      <li><a href="#posteriori">Posteriori</a></li>
      <li><a href="#mmcm">MMCM</a></li>
      <li><a href="#plot">Plot</a></li>
      <li><a href="#relevante-kovariablen">Relevante Kovariablen</a></li>
      <li><a href="#lasso">Lasso</a></li>
      <li><a href="#bayesianischer-lasso">Bayesianischer Lasso</a></li>
      <li><a href="#elastic-net">Elastic Net</a></li>
      </ul>
</li>
      <li>
<a href="#indikatorvariablen">Indikatorvariablen</a><ul class="nav nav-pills nav-stacked">
<li><a href="#indikatorvariablen-1">Indikatorvariablen</a></li>
      <li><a href="#ising-feld">Ising-Feld</a></li>
      </ul>
</li>
      <li>
<a href="#spike-and-slab">Spike and Slab</a><ul class="nav nav-pills nav-stacked">
<li><a href="#spike-and-slab-1">Spike and Slab</a></li>
      <li><a href="#wgrr-und-gen">WGRR und gen</a></li>
      <li><a href="#spikeslabgam-paket">spikeslabGAM paket</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Volker Schmid.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
